tests/integration/tls/test_tls.py

```diff
@@ -154,3 +157,72 @@ async def test_tls_renewal(ops_test: OpsTest) -> None:
         updated_certs["http_certificates_list"][0]["not_before"]
         > current_certs["http_certificates_list"][0]["not_before"]
     )
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_tls_expiration(ops_test: OpsTest) -> None:
+    """Test that expiring TLS certificates are renewed."""
+    # before we can run this test, need to clean up and deploy with different config
+    await ops_test.model.remove_application(APP_NAME, block_until_done=True)
+    await ops_test.model.remove_application(TLS_CERTIFICATES_APP_NAME, block_until_done=True)
+
+    # Deploy TLS Certificates operator
+    config = {"ca-common-name": "CN_CA", "certificate-validity": "1"}
+    await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="stable", config=config)
+    await wait_until(ops_test, apps=[TLS_CERTIFICATES_APP_NAME], apps_statuses=["active"])
+
+    # Deploy Opensearch operator
+    await ops_test.model.set_config(MODEL_CONFIG)
+    my_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(
+        my_charm,
+        num_units=1,
+        series=SERIES,
+    )
+
+    await wait_until(
+        ops_test,
+        apps=[APP_NAME],
+        units_statuses=["blocked"],
+        wait_for_exact_units=1,
+    )
+
+    # Now apply a hack to make the certificate secrets expire in short time
+    # set the secret expiry to a fixed timedelta of 300 seconds to give time to start initially
+    # this happens on the tls_certificates lib and we apply the patch via sed-command
+    search_expression = "expire=self._get_next_secret_expiry_time\\(certificate\\)"
+    replace_expression = "expire=timedelta\\(seconds=180\\)"
+
+    unit_id = get_application_unit_ids(ops_test, APP_NAME)[0]
+    unit_ip = await get_leader_unit_ip(ops_test)
+    lib_file = f"/var/lib/juju/agents/unit-opensearch-{unit_id}/charm/lib/charms/tls_certificates_interface/v3/tls_certificates.py"
+    cmd = f"juju ssh {APP_NAME}/{unit_id} sudo sed -i 's/{search_expression}/{replace_expression}/g' {lib_file}"
+    subprocess.run(cmd, shell=True)
```

Comment:
Huh though it's a pretty cool hack, I'm afraid it strongly wouldn't stand the test of time.

The TLS lib's owners clearly should have the liberty to change the name of their helper functions. Howerver now if that happens, we have a broken pipeline.
(Which is a highly unrewarding experience for the Developer working on a totally unrelated task to debug :-) )

I'm wondering if 
 - there may be a long-term reliable way to add a hack (like monkeypatching a public function like `generate_certificates` or so)

However here I would stop for a moment.

The primary thing we wanted to check was that the notification was raised, no?
I.e. the new behavior introduce in the Opensearch charm code.

Now _that_ could be verified in a reliable way ASSUMING that Juju secret expiry is to be used as a mean of the notification workflow from TLS team. (This we could easily check with them if the secret expiry is an "interface-level" part of their workflow, that should be documented and preserved on their side -- or rather an implementation detail they may change in the future.)

ASSUMING it's part of their public workflow, we could change the secret expiry by resetting the secret info -- which is a safer hack than applying `sed` on someone else's code :-) 

At this point we assume that the rest of the TLS workflow (as shipped by the TLS operator and library) and it's not our role to test if their workflow is safe. Of course if we want to --for the sake of the security of our charm-- then we must find a safe way that could preserve the behavior. I'm actually thinking to ask the TLS libs team to provide a way we could set shorter expiry perhaps than 1d...?

Reply:
I understand your point here, and yes, if they would change that function name, we would have to do work on our side. But given that this in not in their charm but in the libs, the failing test would be the direct impact of a lib update and would thus not cause flaky tests but instead only the test to fail on that PR updating this lib.

I'd like to add that this hack/patch is very helpful for this test to be a "quick" feedback loop. I have to admit that I don't consider waiting for one hour to be acceptable for an integration test, except if there is absolutely no other way around.

Reply:
Several other approaches have been discussed, but none was working. Even directly executing the hook tools of juju, in this case for `secret-set` didn't work out. Therefore I will add a more detailed comment, but leave it as it is for now.

Reply:
All right, after a small research on the topic, if I'm correct this should work:

```
juju exec -u opensearch/4 -- /var/lib/juju/tools/unit-opensearch-4/secret-set cr3ib0gqjdij98lt3h00 --expire <I'm_not_sure_of_the_format>
```
Reverse engineering it from the `ops` library (https://ops.readthedocs.io/en/latest/_modules/ops/model.html -> serach for "def secret_set" )  the format should be `datetime.isoformat()`

Reply:
After running multiple tests with `juju exec ... secret-set` it turned out that this solution will also not be feasible, thus I defaulted back to accepting a one-hour duration of the test-run and adjusted the test for this.